crates/fuel-core/src/combined_database.rs

```diff
@@ -78,6 +78,156 @@ impl CombinedDatabase {
         Ok(())
     }
 
+    #[cfg(feature = "backup")]
+    pub fn backup(
+        db_dir: &std::path::Path,
+        backup_dir: &std::path::Path,
+    ) -> crate::database::Result<()> {
+        use tempfile::TempDir;
+
+        let temp_backup_dir = TempDir::new().map_err(|e| {
+            crate::database::Error::BackupError(anyhow::anyhow!(
+                "Failed to create temporary backup directory: {}",
+                e
+            ))
+        })?;
+
+        let mut must_rollback = false;
```

Comment:
Instead of having `must_rollback` you could have internal private function(that expected `temp_backup_dir` as an input) that do actually backup and returns an error. And in the case of an error, you just remove the dir(but the temp folder will die automatically on `Drop`, so you don't need to do it). If there no error, you can rename folders=)

In this case you also don't need to do loop with `break`

Reply:
addressed in d3d8bffcae